network/p2p/gossip/bloom.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"encoding/binary"
+	"hash"
+	"time"
+
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+	"golang.org/x/exp/rand"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var _ hash.Hash64 = (*hasher)(nil)
+
+// NewBloomFilter returns a new instance of a bloom filter with at most
+// [maxExpectedElements] elements anticipated at any moment, and a false
+// positive probability of [falsePositiveProbability].
+func NewBloomFilter(
+	maxExpectedElements uint64,
+	falsePositiveProbability float64,
+) (*BloomFilter, error) {
+	bloom, err := bloomfilter.NewOptimal(
+		maxExpectedElements,
+		falsePositiveProbability,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	bloomFilter := &BloomFilter{
+		Bloom: bloom,
+		Salt:  randomSalt(),
+	}
+	return bloomFilter, nil
+}
+
+type BloomFilter struct {
+	Bloom *bloomfilter.Filter
+	// Salt is provided to eventually unblock collisions in Bloom. It's possible
+	// that conflicting Gossipable items collide in the bloom filter, so a salt
+	// is generated to eventually resolve collisions.
+	Salt ids.ID
+}
+
+func (b *BloomFilter) Add(gossipable Gossipable) {
+	h := gossipable.GetID()
+	salted := &hasher{
+		hash: h[:],
+		salt: b.Salt,
+	}
+	b.Bloom.Add(salted)
+}
+
+func (b *BloomFilter) Has(gossipable Gossipable) bool {
+	h := gossipable.GetID()
+	salted := &hasher{
+		hash: h[:],
+		salt: b.Salt,
+	}
+	return b.Bloom.Contains(salted)
+}
+
+// ResetBloomFilterIfNeeded resets a bloom filter if it breaches a target false
+// positive probability. Returns true if the bloom filter was reset.
+func ResetBloomFilterIfNeeded(
+	bloomFilter *BloomFilter,
+	falsePositiveProbability float64,
+) bool {
+	if bloomFilter.Bloom.FalsePosititveProbability() < falsePositiveProbability {
+		return false
+	}
+
+	// it's not possible for this to error assuming that the original
+	// bloom filter's parameters were valid
+	fresh, _ := bloomfilter.New(bloomFilter.Bloom.M(), bloomFilter.Bloom.K())
+	bloomFilter.Bloom = fresh
+	bloomFilter.Salt = randomSalt()
+	return true
+}
+
+func randomSalt() ids.ID {
+	salt := ids.ID{}
+	r := rand.New(rand.NewSource(uint64(time.Now().Nanosecond())))
+	_, _ = r.Read(salt[:])
+	return salt
+}
```

Comment:
Could consider making this a utility in the `ids` package. Also, is there a reason we use `"golang.org/x/exp/rand"`? In the rest of the repo we either use `crypto/rand` or `math/rand`

Reply:
Took me a while to remember why I did this. It was because `exp/rand` always returns a `nil` error. I think using `crypto/rand` is going to be better for this though since this is a generic gossip package.